### PR TITLE
Improved performance of Product List page

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | We have more than 160 SQL queries executed on the Product List Page with 7 products, it's TOO much
| Type?         | improvement
| Category?     | BO
| BC breaks?    | yes (an empty function became abstract)
| Deprecations? | no
| How to test?  | The page should work as good as before, but faster

[Blackfire report](https://blackfire.io/profiles/compare/655f244c-af95-4ede-9785-ef567a1c5ac3/graph)

perf_improvements

#### Important guidelines

* Make sure [your local branch is up to date](https://help.github.com/articles/syncing-a-fork/) before commiting your changes!
* Your code MUST respect [our Coding Standards](http://doc.prestashop.com/display/PS16/Coding+Standards) (for code written in PHP, JavaScript, HTML/CSS/Smarty/Twig, SQL)!
* Your commit name MUST respect our [naming convention](http://doc.prestashop.com/display/PS16/How+to+write+a+commit+message)!

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8690)

### CODE REVIEW

File: src/PrestaShopBundle/Security/Admin/EmployeeProvider.php [hunk 1]
```diff
@@ -58,12 +58,21 @@ public function __construct(LegacyContext $context)
      */
     public function loadUserByUsername($username)
     {
-        if (isset($this->legacyContext->employee) && $this->legacyContext->employee->email == $username) {
-            $employee = new Employee($this->legacyContext->employee);
-            $employee->setRoles(
-                Access::getRoles($this->legacyContext->employee->id_profile)
-            );
-            return $employee;
+        static $employees = array();
```
Comment (Quetzacoalt91) [inline src/PrestaShopBundle/Security/Admin/EmployeeProvider.php]: And what about a class property to store the cache (which can be reset with another function)? I remember the issues you got with @tomlev :)

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Security/Admin/EmployeeProvider.php]: nope, there is better: using the cache system from Symfony. But I don't know yet if it's the right time to introduce it as I don't have time to introduce it almost everywhere + document it .

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Security/Admin/EmployeeProvider.php]: as soon as my build is green, I'll try to introduce Symfony cache for this class *only* in a specific commit.

**Comment by eternoendless** [comment]
Nice one! :clap: 

---

Reviewed 3 of 5 files at r1, 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[classes/Access.php, line 122 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L3Y0cJ1ljbgiCrL-41V:-L3Y0cJ1ljbgiCrL-41W:bb5hhke) ([raw file](https://github.com/prestashop/prestashop/blob/95784fe05651f1246ce4bfcda9cb3b63a72a075d/classes/Access.php#L122)):*
> ```PHP
>                 FROM `'._DB_PREFIX_.'module_access` ma 
>                 WHERE ma.`id_profile` = "'.$idProfile.'"
>             )
> ```

Wouldn't this be simpler?

```sql
SELECT r.`slug`
FROM `'._DB_PREFIX_.'authorization_role` r
INNER JOIN `'._DB_PREFIX_.'module_access` ma ON ma.`id_authorization_role` = r.`id_authorization_role`
WHERE ma.`id_profile` = "'.$idProfile.'"
```

---

*[src/Adapter/Product/AdminProductDataProvider.php, line 82 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L3Y39Eboc3LlpKw1p5D:-L3Y39Eboc3LlpKw1p5E:bn0bamw) ([raw file](https://github.com/prestashop/prestashop/blob/95784fe05651f1246ce4bfcda9cb3b63a72a075d/src/Adapter/Product/AdminProductDataProvider.php#L82)):*
> ```PHP
>     public function getPersistedFilterParameters()
>     {
>         static $filters = null;
> ```

This cache should be cleanable

---

*[src/PrestaShopBundle/Security/Admin/EmployeeProvider.php, line 61 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L3HrZa5bF8G3ujLRwX7-r1-61:-L3Y3WNMdRbEF4DJ0zco:b-876jpc) ([raw file](https://github.com/prestashop/prestashop/blob/7c42189c2390d04b6039aa414a3648b4d5f4e171/src/PrestaShopBundle/Security/Admin/EmployeeProvider.php#L61)):*
<details><summary><i>Previously, mickaelandrieu (Mickaël Andrieu) wrote…</i></summary><blockquote>

as soon as my build is green, I'll try to introduce Symfony cache for this class *only* in a specific commit.
</blockquote></details>

Ok but that's a blocker though :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-:-L3Y3rKsABDkhZuK_ux1:b2jnje9)*

**Comment by mickaelandrieu** [comment]
> Wouldn't this be simpler?

Maybe, but is it more performant? Provide a graph please :)

> This cache should be cleanable

> Ok but that's a blocker though :)

Yes, it's on my free time so we need to be patient ^^

**Comment by eternoendless** [comment]
Reviewed 1 of 1 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[classes/Access.php, line 122 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L3Y0cJ1ljbgiCrL-41V:-L3mvAT6Qun0oRqmZ4Qu:b-njxeh) ([raw file](https://github.com/prestashop/prestashop/blob/95784fe05651f1246ce4bfcda9cb3b63a72a075d/classes/Access.php#L122)):*

> Maybe, but is it more performant? Provide a graph please :)

I'm under the impression that a `JOIN` is better than an `IN ( <subquery> )`, but looks like MariaDB is smart enough to optimize it on its own to the same query (I'm not sure about MySQL... @jocel1?):

```sql
explain SELECT r.`slug`
FROM `ps_authorization_role` r
INNER JOIN `ps_module_access` ma
ON ma.`id_authorization_role` = r.`id_authorization_role`
WHERE ma.`id_profile` = 1;
```

| id | select_type | table | type   | possible_keys | key     | key_len | ref                             | rows | Extra       | 
|----|-------------|-------|--------|---------------|---------|---------|---------------------------------|------|-------------| 
| 1  | SIMPLE      | ma    | ref    | PRIMARY       | PRIMARY | 4       | const                           | 216  | Using index | 
| 1  | SIMPLE      | r     | eq_ref | PRIMARY       | PRIMARY | 4       | ps173x.ma.id_authorization_role | 1    |             | 

```sql
explain SELECT r.`slug`
FROM `ps_authorization_role` r
WHERE r.`id_authorization_role` IN (
    SELECT ma.`id_authorization_role`
    FROM `ps_module_access` ma
    WHERE ma.`id_profile` = 1
)
```

| id | select_type | table | type   | possible_keys | key     | key_len | ref                             | rows | Extra       | 
|----|-------------|-------|--------|---------------|---------|---------|---------------------------------|------|-------------| 
| 1  | PRIMARY     | ma    | ref    | PRIMARY       | PRIMARY | 4       | const                           | 216  | Using index | 
| 1  | PRIMARY     | r     | eq_ref | PRIMARY       | PRIMARY | 4       | ps173x.ma.id_authorization_role | 1    |             | 

Run on MariaDB 10.2.6

--

Anyway, personally, I find the JOIN syntax easier to understand.

---

*[config/defines.inc.php, line 29 at r4](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L3mybrghDJY4hbMrs2L:-L3mybrghDJY4hbMrs2M:b-t8hhw6) ([raw file](https://github.com/prestashop/prestashop/blob/63eeabd602df78f7270f798b8948c75d6d6f7192/config/defines.inc.php#L29)):*
> ```PHP
> /* Debug only */
> if (!defined('_PS_MODE_DEV_')) {
>     define('_PS_MODE_DEV_', false);
> ```

Should be kept to `true`

---

*[src/PrestaShopBundle/Security/Admin/EmployeeProvider.php, line 61 at r1](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L3HrZa5bF8G3ujLRwX7-r1-61:-L3mzWBYoonboPFEy4M1:bcvkx4p) ([raw file](https://github.com/prestashop/prestashop/blob/7c42189c2390d04b6039aa414a3648b4d5f4e171/src/PrestaShopBundle/Security/Admin/EmployeeProvider.php#L61)):*
> Yes, it's on my free time so we need to be patient ^^

Sure, no rush! :sunglasses: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8690)*

**Comment by jocel1** [comment]
*[classes/Access.php, line 122 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L3Y0cJ1ljbgiCrL-41V:-L3mzygysLCOUtYTFrmG:buar6xs) ([raw file](https://github.com/prestashop/prestashop/blob/95784fe05651f1246ce4bfcda9cb3b63a72a075d/classes/Access.php#L122)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>


> Maybe, but is it more performant? Provide a graph please :)

I'm under the impression that a `JOIN` is better than an `IN ( <subquery> )`, but looks like MariaDB is smart enough to optimize it on its own to the same query (I'm not sure about MySQL... @jocel1?):

```sql
explain SELECT r.`slug`
FROM `ps_authorization_role` r
INNER JOIN `ps_module_access` ma
ON ma.`id_authorization_role` = r.`id_authorization_role`
WHERE ma.`id_profile` = 1;
```

| id | select_type | table | type   | possible_keys | key     | key_len | ref                             | rows | Extra       | 
|----|-------------|-------|--------|---------------|---------|---------|---------------------------------|------|-------------| 
| 1  | SIMPLE      | ma    | ref    | PRIMARY       | PRIMARY | 4       | const                           | 216  | Using index | 
| 1  | SIMPLE      | r     | eq_ref | PRIMARY       | PRIMARY | 4       | ps173x.ma.id_authorization_role | 1    |             | 

```sql
explain SELECT r.`slug`
FROM `ps_authorization_role` r
WHERE r.`id_authorization_role` IN (
    SELECT ma.`id_authorization_role`
    FROM `ps_module_access` ma
    WHERE ma.`id_profile` = 1
)
```

| id | select_type | table | type   | possible_keys | key     | key_len | ref                             | rows | Extra       | 
|----|-------------|-------|--------|---------------|---------|---------|---------------------------------|------|-------------| 
| 1  | PRIMARY     | ma    | ref    | PRIMARY       | PRIMARY | 4       | const                           | 216  | Using index | 
| 1  | PRIMARY     | r     | eq_ref | PRIMARY       | PRIMARY | 4       | ps173x.ma.id_authorization_role | 1    |             | 

Run on MariaDB 10.2.6

--

Anyway, personally, I find the JOIN syntax easier to understand.
</blockquote></details>

Always prefer JOIN over IN (subquery) if you can, earlier version of MySQL are not able to optimise them efficiently (but MariaDB and recent version of MySQL are able to optimize them)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8690)*

**Comment by mickaelandrieu** [comment]
Even if build is green, the pr is not yet finished: as default cache of Symfony is file system and not static array, I need to declare a new adapter.

**Comment by mickaelandrieu** [comment]
I've used the Symfony static cache, so we can rely on Symfony services/commands to clear the cache if needed. I've also updated the link of the Blackfire graph.

This time, I've finished the first bash of improvements: thanks Blackfire :angel:

File: src/Adapter/Product/AdminProductDataProvider.php [hunk 2]
```diff
@@ -80,19 +83,32 @@ public function __construct(EntityManager $entityManager, ImageManager $imageMan
...
-        $filter = $this->entityManager->getRepository('PrestaShopBundle:AdminFilter')->findOneBy(array(
-            'employee' => $employee->id ?: 0,
-            'shop' => $shop->id ?: 0,
-            'controller' => 'ProductController',
-            'action' => 'catalogAction',
-        ));
-        /* @var $filter AdminFilter */
-        if (!$filter) {
-            return AdminFilter::getProductCatalogEmptyFilter();
+        $employeeId = $employee->id ?: 0;
```
Comment (Quetzacoalt91) [inline src/Adapter/Product/AdminProductDataProvider.php:87]: For people wondering what is the compatibility for a shorthand ternary test: PHP 5.3 required.

**Review by Quetzacoalt91 — COMMENTED** [review]
We have a implementation of the Symfony Cache based on PSR-6 here.

Does the `app.` in your cache keys make the tags available?

**Comment by eternoendless** [comment]
Reviewed 4 of 6 files at r5, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[classes/controller/Controller.php, line 648 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L40OXpaT_3utAlwo_ec:-L40OXpaT_3utAlwo_ed:b-gc7kxq) ([raw file](https://github.com/prestashop/prestashop/blob/8bb2066d0d6f749832bc82849ec337671d217aa0/classes/controller/Controller.php#L648)):*
> ```PHP
>      * @return Symfony\Component\DependencyInjection\ContainerBuilder
>      */
>     abstract protected function buildContainer();
> ```

This is BC break in my opinion

---

*[src/PrestaShopBundle/Security/Admin/EmployeeProvider.php, line 68 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L40PEns8AXuuoboXhoL:-L40PEns8AXuuoboXhoM:bdocf7y) ([raw file](https://github.com/prestashop/prestashop/blob/8bb2066d0d6f749832bc82849ec337671d217aa0/src/PrestaShopBundle/Security/Admin/EmployeeProvider.php#L68)):*
> ```PHP
>     public function loadUserByUsername($username)
>     {
>         $cacheKey = sha1($username);
> ```

Why the sha1?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8690)*

**Comment by Quetzacoalt91** [comment]
*[classes/controller/Controller.php, line 648 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L40OXpaT_3utAlwo_ec:-L40SgTY7y30pRX6Lb_9:b5dy9d4) ([raw file](https://github.com/prestashop/prestashop/blob/8bb2066d0d6f749832bc82849ec337671d217aa0/classes/controller/Controller.php#L648)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

This is BC break in my opinion
</blockquote></details>

As long as AdminController & FrontController declared the method, I'm confident is saying this won't break the compatibility.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8690)*

**Comment by eternoendless** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[classes/controller/Controller.php, line 648 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L40OXpaT_3utAlwo_ec:-L40WYzvgc9EJga21gk4:b-y12tum) ([raw file](https://github.com/prestashop/prestashop/blob/8bb2066d0d6f749832bc82849ec337671d217aa0/classes/controller/Controller.php#L648)):*
<details><summary><i>Previously, Quetzacoalt91 (Thomas N) wrote…</i></summary><blockquote>

As long as AdminController & FrontController declared the method, I'm confident is saying this won't break the compatibility.
</blockquote></details>

Still needs to be documented in case people extend Controller directly

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8690)*

**Comment by Quetzacoalt91** [comment]
By the way, I used the cache system provided by Doctrine. If we use the system coming from Symfony, we will have to replace the old one.

File: src/Adapter/Product/AdminProductDataProvider.php (hunk 2 above)
Comment (mickaelandrieu) [reply src/Adapter/Product/AdminProductDataProvider.php]: not in develop branch my friend, we are using PHP 5.6

**Comment by mickaelandrieu** [comment]
> This is BC break in my opinion

I'm in favor of document it, but where? We don't have a changelog, didn't we?

> Why the sha1?

Cache component doesn't like spaces and special characters like `@`.

> By the way, I used the cache system provided by Doctrine. If we use the system coming from Symfony, we will have to replace the old one.

Except if I'm wrong, Cache component provides Doctrine support: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Cache/Adapter/DoctrineAdapter.php.

I see no hurry here for this pull request. You asked me to implement something more reliable than static PHP cache and I did it. Introducing Symfony Cache component everywhere in PrestaShop is a massive task where we need to define:

- what needs to be cached;
- how many time: during a request, during 1 hour or during 1 week?

For each page of PrestaShop. Looking at the current roadmap, this may be planned and released in 1.7.5.

**Comment by eternoendless** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[classes/controller/Controller.php, line 648 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L40OXpaT_3utAlwo_ec:-L4l2jd_esCtS3z7wH1Q:b44uped) ([raw file](https://github.com/prestashop/prestashop/blob/8bb2066d0d6f749832bc82849ec337671d217aa0/classes/controller/Controller.php#L648)):*
> I'm in favor of document it, but where? We don't have a changelog, didn't we?

You can put it here:



---

*[src/PrestaShopBundle/Security/Admin/EmployeeProvider.php, line 68 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/8690#-L40PEns8AXuuoboXhoL:-L4l2FZzRB9Yxp5elzIT:b-bcbt2e) ([raw file](https://github.com/prestashop/prestashop/blob/8bb2066d0d6f749832bc82849ec337671d217aa0/src/PrestaShopBundle/Security/Admin/EmployeeProvider.php#L68)):*
> Cache component doesn't like spaces and special characters like `@`.

Ok, please add a comment explaining that so that other people will know :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8690)*

**Comment by mickaelandrieu** [comment]
> We have a implementation of the Symfony Cache based on PSR-6 here.
   Does the app. in your cache keys make the tags available?

Good question: regarding the type hinted interface, the answer is no. But as `TagAwareAdapterInterface` extends the previous one (in the end), it may be the case.

Before introduce tags, we need a strategy for caching in PrestaShop: introducing tags later won't break the API :+1:

**Comment by mickaelandrieu** [comment]
ping @eternoendless, this PR was already reviewed but it was too risky to be merged in 1.7.4.0 I guess.

Anyway, this is now up to date with `develop` branch :)

**Comment by mickaelandrieu** [comment]
ping @eternoendless, this PR was already reviewed but it was too risky to be merged in 1.7.4.0. Let's make them available in develop branch asap to prevent bugs if introduced, don't you think?

Anyway, this is now up to date with develop branch :)

**Review by eternoendless — APPROVED** [review]
> Reviewed 9 of 10 files at r7.
> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8690)* status: all files reviewed, 1 unresolved discussion (waiting on @Quetzacoalt91)

___
*[src/PrestaShopBundle/Security/Admin/EmployeeProvider.php, line 59 at r7](https://reviewable.io/reviews/prestashop/prestashop/8690#-LJEmijH6r9Y9-bEPemm:-LJEmijH6r9Y9-bEPemn:b7o0v8u) ([raw file](https://github.com/prestashop/prestashop/blob/b95d66cf9891fe05eb0798089d0dd360b25593f3/src/PrestaShopBundle/Security/Admin/EmployeeProvider.php#L59)):*
> ```PHP
>     /**
>      * Fetch the Employee entity that matches the given username.
>      * Cache system doesn't supports "@" character, so we rely on a sha1 expression.
> ```

This comment is about implementation, so it should go in the appropriate line (just above `sha1(...)`)

**Comment by PierreRambaud** [comment]
Thanks @mickaelandrieu and everyone who works on it.

**Pull request merged**